### PR TITLE
Find mergeable PRs to trigger batch tests for.

### PR DESCRIPTION
**Pull request by spxtr**
This copies some logic from cmd/splice.

/area prow
/assign @rmmh

### CODE REVIEW

File: prow/tide/tide.go [hunk 1]
```diff
@@ -234,6 +237,38 @@ func accumulate(presubmits []string, prs []pullRequest, pjs []kube.ProwJob) (suc
...
+	}
+	return nums
+}
+
+func (c *Controller) pickBatch(sp subpool) ([]pullRequest, error) {
+	r, err := c.gc.Clone(sp.org + "/" + sp.repo)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Clean()
```
Comment (rmmh) [inline prow/tide/tide.go]: TODO: don't erase the whole repo

File: prow/tide/tide.go [hunk 2]
```diff
@@ -234,6 +237,38 @@ func accumulate(presubmits []string, prs []pullRequest, pjs []kube.ProwJob) (suc
...
+	if err := r.Checkout(sp.sha); err != nil {
+		return nil, err
+	}
+	var res []pullRequest
+	for _, pr := range sp.prs {
+		// TODO(spxtr): Check the actual statuses for individual jobs.
+		if string(pr.Commits.Nodes[0].Commit.Status.State) != "SUCCESS" {
+			continue
+		}
+		if ok, err := r.Merge(string(pr.HeadRef.Target.OID)); err != nil {
```
Comment (rmmh) [inline prow/tide/tide.go]: Don't you need to fetch the PR heads?

Comment (spxtr) [reply prow/tide/tide.go]: Nope, it's a mirror clone so it has all that stuff.

File: prow/tide/tide.go (hunk 1 above)
Comment (spxtr) [reply prow/tide/tide.go]: It's fine, the git client keeps a mirror clone that we checkout from. A git checkout from a local mirror clone is fast.

File: prow/cmd/tide/main.go [hunk 3]
```diff
@@ -78,7 +79,14 @@ func main() {
...
-	c := tide.NewController(logrus.WithField("controller", "tide"), ghc, kc, configAgent)
+	gc, err := git.NewClient()
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting git client.")
+	}
+	defer gc.Clean()
+	gc.Logger = logrus.WithField("client", "git")
+
+	c := tide.NewController(logrus.WithField("controller", "tide"), ghc, kc, configAgent, gc)
 	if err != nil {
```
Comment (kargakis) [inline prow/cmd/tide/main.go]: redundant check

**Comment by spxtr** [comment]
Adding tests..

File: prow/cmd/tide/main.go (hunk 3 above)
Comment (spxtr) [reply prow/cmd/tide/main.go]: Done.

**Comment by spxtr** [comment]
@rmmh ptal, should be good to go in now.

File: prow/tide/tide.go [hunk 4]
```diff
@@ -234,6 +237,45 @@ func accumulate(presubmits []string, prs []pullRequest, pjs []kube.ProwJob) (suc
...
 
+func prNumbers(prs []pullRequest) []int {
+	var nums []int
+	for _, pr := range prs {
+		nums = append(nums, int(pr.Number))
+	}
+	return nums
+}
+
+func (c *Controller) pickBatch(sp subpool) ([]pullRequest, error) {
```
Comment (kargakis) [inline prow/tide/tide.go]: How are we going to ensure that the single PR which we have triggered tests for is going to be included here?

Comment (spxtr) [reply prow/tide/tide.go]: It's not guaranteed to be here currently. The reason why we might want it to be is that it gives it a second chance to dodge a flake, but I'm not sure how important that is. This does mean that a PR might be retested at the same time as the batch, pass tests, but not merge due to the batch taking priority.

If we settle on a testing order then we can ensure that the single PR is in the batch pretty easily. Currently the submit queue uses least-recently-lgtm'd, but we can also go with oldest PR or something.

Comment (kargakis) [reply prow/tide/tide.go]: Would be nice to capture into an issue for more discussion.

Comment (spxtr) [reply prow/tide/tide.go]: There's some discussion in the design doc, but it's mostly marked as resolved.

**Pull request merged**